### ISSUE TITLE
Variables are being coerced into strings by apollo-client in some situations

### ISSUE THREAD

**Issue by a-s-o** [apollostack/apollo-client#483]
Hi,

I am noticing that the variables passed to a query are being coerced. I would like to know whether this is intended behavior or it is an error? Here is an example:

```javascript
    const query = gql`
      query authorDetails ($id: String) {
        author(id: $id) {
          name
          books {
            id
            title
          }
        }
      }
    `;

    client
      .query({ query, variables: { id: [1, 2, undefined, null, {}] } })
      .then(handleSuccess)
      .catch(logError);
```

In the above case, my resolver on the backend receives args of `{ id: '1,2,,,[object Object]' }`. Not sure if it is this package or `apollo-server` that is doing coercing the id to string. I would normally expect something like this to throw an error.

I would appreciate if someone can clear this for me. Thanks for the great package.

**Comment by Poincare** [comment]
I don't think doing type checking on the variable values should be Apollo Client's responsibility - it should simply follow the query structure.

**Comment by a-s-o** [comment]
Fair, we can leave it at that. However, this is a desireable check. I can even do the check on the application side but the issue is that after the coersion I have less information about the incoming parameter. This I think is a bad default. So i am enquiring which package is doing that?

**Comment by stubailo** [comment]
I think this is something your server is doing. Apollo client simply JSON-serializes the variables and passes them down, which shouldn't be doing any coercion.

Are you using grpahql-js?

**Comment by a-s-o** [comment]
I'm using apollo-server with express

**Comment by stubailo** [comment]
I think I'd be surprised if this was done by apollo client. Have you tried putting in similar variables in GraphiQL and seeing if the same thing happens?

**Comment by a-s-o** [comment]
Okay, i can check that and report back in the evening. Thanks for the suggestion.

**Comment by a-s-o** [comment]
Hi, reporting back. The issue does not occur through graphiql.

Here is my query in graphiql
```graphql
{
  author(id: [1, 2, undefined, {}]) {
    name
    books {
      id
      title
    }
  }
}
```
The response is:
```json
{
  "errors": [
    {
      "message": "Argument \"id\" has invalid value [1, 2, undefined, {}].\nExpected type \"String\", found [1, 2, undefined, {}].",
      "locations": [
        {
          "line": 2,
          "column": 14
        }
      ]
    }
  ]
}
```

Note: I skipped the null value from my original example above as that causes a SyntaxError in graphiql.

Thanks for taking a look at this.

**Comment by stubailo** [comment]
@a-s-o wait a second, this doesn't quite reproduce. Can you pass the `id` in the `variables` panel of GraphiQL? Instead of inline in the query? And see if it still works the same.

**Comment by Slava** [comment]
One of the solutions that we can implement here, is a quick check of variables, according to their declared type. If the type is something simple like `String!` or `Int!`, it is fairly easy to check.

**Comment by Slava** [comment]
We've done some testing with other GraphQL servers available and it seems to be the problem of the graphql reference implementation? Here is a link that demonstrates the same problem: [link](https://www.graphqlhub.com/playground?query=%23%20Welcome%20to%20GraphQLHub!%20Type%20your%20GraphQL%20query%20here%2C%20or%0A%23%20explore%20the%20%22Docs%22%20to%20the%20right%0A%0Aquery%20(%24asdf%3AString!)%7B%0A%20%20hn2%20%7B%0A%20%20%20%20nodeFromHnId(id%3A%24asdf%2C%20isUserId%3Atrue)%20%7B%0A%20%20%20%20%20%20id%0A%20%20%20%20%7D%0A%20%20%7D%0A%7D&variables=%7B%0A%20%20%22asdf%22%3A%20%5B1%2C2%5D%0A%7D). Here is another link with a pad that demonstrates correct type checking in case of an inline array: [link](https://www.graphqlhub.com/playground?query=%23%20Welcome%20to%20GraphQLHub!%20Type%20your%20GraphQL%20query%20here%2C%20or%0A%23%20explore%20the%20%22Docs%22%20to%20the%20right%0A%0Aquery%7B%0A%20%20hn2%20%7B%0A%20%20%20%20nodeFromHnId(id%3A%5B1%2C2%2C3%5D%2C%20isUserId%3Atrue)%20%7B%0A%20%20%20%20%20%20id%0A%20%20%20%20%7D%0A%20%20%7D%0A%7D&variables=%7B%0A%20%20%22asdf%22%3A%20%5B1%2C2%5D%0A%7D).

Type checking is not a responsibility of the client. Although, I have implemented some simple type checking in #522, we are thinking twice before merging it to master.

**Comment by stubailo** [comment]
This is confirmed as an issue with GraphQL-JS itself, and therefore an issue with any JavaScript GraphQL server based on it; read below.

[Here's the file](https://github.com/graphql/graphql-js/blob/08a00aa4eb2a58c1aece167ee70774fae7c5393c/src/type/scalars.js) where the parsing for scalar types is defined. You can see:

- [parseValue](https://github.com/graphql/graphql-js/blob/08a00aa4eb2a58c1aece167ee70774fae7c5393c/src/type/scalars.js#L82) simply coerces to a string no matter what, this is what is used for variables
- [parseLiteral](https://github.com/graphql/graphql-js/blob/08a00aa4eb2a58c1aece167ee70774fae7c5393c/src/type/scalars.js#L83) actually checks the type of the AST node, this is used for the inline arguments

I'd say we should report this issue on graphql-js, and perhaps submit a PR, since it's definitely unexpected to me that there is no validation at all on strings or numbers passed via variables.
